### PR TITLE
aws_lambda filter in json mode should use content-type header from json response

### LINKED ISSUE #16615: aws-lambda passing wrong content-type when  payload_passthrough=false
*Description*:
>When returning the response from a Lambda function to the client Envoy sets the 'content-type' header to 'application/json' instead of the content-type set by the Lambda.  It seems the aws-lambda extension is pulling the content-type from the aws lambda api connection (line 351 in aws_lambda_filter.cc)

*Repro steps*:
> Create a Lambda function that responds with content-type = 'text/html' or anything other than 'application/json'.  Enable the aws-lambda extension with payload_passthrough = false.  

*Config*:
```
admin:
  address:
    socket_address: { address: 127.0.0.1, port_value: 9901 }
static_resources:
  listeners:
  - address:
      socket_address: { address: 0.0.0.0, port_value: 443 }
    filter_chains:
    - transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
          common_tls_context:
            tls_certificates:
            - certificate_chain:
                filename: "/etc/xx/xx/default.crt"
              private_key:
                filename: "/etc/x/xx/default.key"
      filters:
      - name: envoy.extenstions.filters.network.http_connection_manager.v3
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: AUTO
          stat_prefix: http
          use_remote_address: true
          server_header_transformation: OVERWRITE
          strip_any_host_port: true
          http2_protocol_options:
            max_concurrent_streams: 100
          access_log:
            name: envoy.file_access_log
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
              path: /var/log/envoy/access.log
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match: { prefix: "/" }
                route: { cluster: lambda_egress_gateway }
          http_filters:
          - name: envoy.filters.http.compressor
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.compressor.v3.Compressor
              response_direction_config:
                common_config:
                  min_content_length: 100
                  content_type:
                    - text/html
                    - application/json
                disable_on_etag_header: true
              request_direction_config:
                common_config:
                  enabled:
                    default_value: false
                    runtime_key: request_compressor_enabled
              compressor_library:
                name: text_optimized
                typed_config:
                  "@type": type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip
                  memory_level: 3
                  window_bits: 10
                  compression_level: best_compression
                  compression_strategy: default_strategy
          - name: envoy.filters.http.aws_lambda
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.Config
              arn: "arn:aws:lambda:us-west-1:062333172176:function:xxxx-myfunction-xxxx"
              payload_passthrough: false 
          - name: envoy.filters.http.router
  clusters:
  - name: lambda_egress_gateway
    connect_timeout: 0.55s
    type: LOGICAL_DNS
    lb_policy: ROUND_ROBIN
    metadata:
      filter_metadata:
        com.amazonaws.lambda:
          egress_gateway: true
    load_assignment:
      cluster_name: lambda_egress_gateway
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: lambda.us-west-1.amazonaws.com
                port_value: 443
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
        sni: "*.amazonaws.com"
```

*Logs*:
>[envoy_aws_lambda_content_type_bug.txt](https://github.com/envoyproxy/envoy/files/6523283/envoy_aws_lambda_content_type_bug.txt)

### PR DESCRIPTION
**Pull request by jstewmon**
Commit Message: aws_lambda filter in json mode should use content-type header from json response
Additional Description: fixes #16615 
Risk Level: Low
Testing: unit tested
Docs Changes: None - the new behavior conforms to the existing docs.
Release Notes: Bug fix mentioned in version history
Platform Specific Features: None
Fixes #16615

### CODE REVIEW

**Comment by jstewmon** [comment]
@rgs1 would you mind giving this small PR a review?

File: test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc [hunk 1]
```diff
@@ -531,6 +532,40 @@ TEST_F(AwsLambdaFilterTest, EncodeDataJsonModeTransformToHttp) {
   EXPECT_THAT(cookies, ElementsAre("session-id=42; Secure; HttpOnly", "user=joe"));
 }
 
+/**
+ * encodeData() data in JSON mode should use respect content-type header.
```
Comment (rgs1) [inline test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc:536]: nit: s/use//

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 2]
```diff
@@ -333,6 +333,10 @@ void Filter::dejsonizeResponse(Http::ResponseHeaderMap& headers, const Buffer::I
     return;
   }
 
+  // Use JSON as the default content-type. If the JSON response headers has a
+  // content-type key, it will override.
```
Comment (rgs1) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:337]: nit: If the response headers have a different content-type set, that will be used instead.

**Review by rgs1 — APPROVED** [review]
LGTM modulo some nits – over to @tonya11en for final review & merging. Thanks!

**Comment by marcomagdy** [comment]
LGTM as well FWIW :)

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -38,6 +38,7 @@ Bug Fixes
 ---------
 *Changes expected to improve the state of the world and are unlikely to have negative effects*
 
+* aws_lambda: in JSON mode, the downstream response content-type header will now be set from the content-type entry in the JSON response's headers map, if present.
```
Comment (marcomagdy) [inline docs/root/version_history/current.rst:41]: nit: I don't think there's a "JSON mode" per se. It's either normal or passthrough.
So, this should probably read "unless passthrough is on, the downstream response..."

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**